diff --git a/include/sbi/sbi_domain_context.h b/include/sbi/sbi_domain_context.h
index 31a3a7f..e0f4bda 100644
--- a/include/sbi/sbi_domain_context.h
+++ b/include/sbi/sbi_domain_context.h
@@ -38,4 +38,7 @@ int sbi_domain_context_init(void);
 /* Deinitialize domain context support */
 void sbi_domain_context_deinit(void);

+/** Marks a domain as active without performing a context switch */
+int sbi_domain_change_active(u32 domain_id);
+
 #endif // __SBI_DOMAIN_CONTEXT_H__
diff --git a/lib/sbi/sbi_domain_context.c b/lib/sbi/sbi_domain_context.c
index 407c0d5..5d33cd5 100644
--- a/lib/sbi/sbi_domain_context.c
+++ b/lib/sbi/sbi_domain_context.c
@@ -239,3 +239,33 @@ void sbi_domain_context_deinit(void)
 {
 	sbi_domain_unregister_data(&dcpriv);
 }
+
+int sbi_domain_change_active(u32 target_id)
+{
+	u32 hartindex = current_hartindex();
+	struct sbi_domain *target_dom = 0;
+	struct sbi_domain *current_dom = sbi_domain_thishart_ptr();
+
+	/* Search for the domain in the domain_list */
+	sbi_domain_for_each(target_dom) {
+		if (target_dom->index == target_id)
+			break;
+	}
+
+	// No domain with the defined id
+	if (!target_dom)
+		return SBI_EINVAL;
+
+	/* Assign current hart to target domain */
+	spin_lock(&current_dom->assigned_harts_lock);
+	sbi_hartmask_clear_hartindex(hartindex, &current_dom->assigned_harts);
+	spin_unlock(&current_dom->assigned_harts_lock);
+
+	sbi_update_hartindex_to_domain(hartindex, target_dom);
+
+	spin_lock(&target_dom->assigned_harts_lock);
+	sbi_hartmask_set_hartindex(hartindex, &target_dom->assigned_harts);
+	spin_unlock(&target_dom->assigned_harts_lock);
+
+	return 0;
+}
